README.md
# NotifyModuleV2
The branch is for Community Driven updates.

# How to Install
## To install NotifyV1 Community Version:
```
1: Get the Notify Installer here: "https://www.roblox.com/library/13659358957/"
2: install the Updater in studio, and just click install (it will also update)
3: thats it, just follow the usage guide for details on how to use
```
## To install NotifyV1 Community Version Without Plugin:
```
1: Get the Notify Instance here: "https://www.roblox.com/library/13659448082/"
2: Delete or Move "NotifyUpdater" to "ServerScriptService", the plugin will replace it anyway when installing
3: thats it, just follow the usage guide for details on how to use
```

# Basic usage Guide
Still Writing this section!!




# Features added changed/added:
~Custom colors can be used by sending a table of color values like so:
```
  local notify = require(game.Workspace:WaitForChild("Services"):FindFirstChild("notify")) -- Replace "notify" with the location of the module
  --local color = {["R"]=255, ["G"]=0, ["B"]=0,}
  notify.notify(Player,"TitleCard", "Message", color, DLY)
```

along with that the Tween service was rewritten, so now upon providing ANY delay Notify will use a seperate script to Tween
Allowing the main script(Notify Module) to run without a delay, rather than waiting for the tween service to finish fully

The reason for the tween rewrite was during a Playeradded function would take ages to finish because it has to finish the tween for every player
This workaround allows anyone that uses this verison of notify to call 6 notifactions side-by-side and still run the scripts like datastore loading, or Charachter loading

Original.MD will contain the original Code from simplyNotifiedV1 (Made by @Starnamics)
Changes.MD will contain what i changes i have made in simplyNotifiedV1


# Q&A Section below

# Why this branch exist?
```
For new users intrested in learning coding,
while maintaining and forging a new path for NotifyV1, along with Support in the issue tab.
Examples are provided to help prevent User error.
```
# Where can i Download notify?
```
The original Author has stopped working on NotifyV1, that version is broken as well(for me at least)
in the "CommuintyRelease" branch contains all you need to get started, including an auto-update system
```
# Copyright concerns?
```
1: These Assets provides no majour value over notifyV2 or V3, its simply being kept up-to-date. with active support and community Collabs
2: Any material made by myself does NOT reflect @Starnamics Ability to program or Maintain NotifyV1
3: These scripts/assets are subject to change are open source to others
```
# Can i help build NotifyV1? or Create new assets?
```
1: Yes, but ANY change made to any branch must be reviewed by me
2: Yes, as long as your Asset is Open source, no owner can be assigned to it. otherwise i cant use it
```
# what about exploits or backdooring?
```
1: All Changes are reviewed by myself and others
2: All Changes are tested in a "Clean-Room" to prevent backdooring on any main game
3: if any detections are in place for studio we will not accept that change
```



Changes.md
```
local module = {}
local TweenService = game.Workspace:WaitForChild("Services"):FindFirstChild("notify"):FindFirstChild("TweenService"):FindFirstChild("Event") -- Replace "notify" with the location of the module
local GUIBuilder = require(script:WaitForChild("GUIBuilder"))
local NVersion=script:WaitForChild("Version")
NVersion.Value="V1.05"
print("Started notify Module")
function insertContainer(p)
	if p == nil then
		--Rewrite
		if game:GetService("StarterGui"):WaitForChild("NotifyModuleContainer")==nil then
			GUIBuilder.CreateBin(game:GetService("StarterGui"))
		end
	else
		--Rewrite
		if p:WaitForChild("PlayerGui"):WaitForChild("NotifyModuleContainer")==nil then
			GUIBuilder.CreateBin(p:FindFirstChild("PlayerGui"))
		end
	end
end

insertContainer()

--Preset colors
local colors = {
	NILColor  = {
		["BackdropColor"]=Color3.fromRGB(40, 40, 40),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(40, 40, 40),
	},
	alert  = {
		["BackdropColor"]=Color3.fromRGB(254, 104, 104),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(254, 104, 104),
	},
	success = {
		["BackdropColor"]=Color3.fromRGB(104, 254, 104),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(104, 254, 104),
	},
	warning  = {
		["BackdropColor"]=Color3.fromRGB(254, 254, 104),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(254, 254, 104),
	},
	info = {
		["BackdropColor"]=Color3.fromRGB(104, 104, 254),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(104, 104, 254),
	},
	default = {
		["BackdropColor"]=Color3.fromRGB(254, 254, 254),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(254, 254, 254),
	},
}

--Color functions
function SetColorCus(color)
	local selectedColor
	selectedColor = {
		["BackdropColor"]=Color3.fromRGB(math.clamp(color.BackdropColor.R*255,0,255), math.clamp(color.BackdropColor.G*255,0,255), math.clamp(color.BackdropColor.B*255,0,255)),
		["IconboxColor"]=Color3.fromRGB(math.clamp(color.IconboxColor.R*255,0,255), math.clamp(color.IconboxColor.G*255,0,255), math.clamp(color.IconboxColor.B*255,0,255)),
		["MSGBoxColor"]=Color3.fromRGB(math.clamp(color.MSGBoxColor.R*255,0,255), math.clamp(color.MSGBoxColor.G*255,0,255), math.clamp(color.MSGBoxColor.B*255,0,255)),
		["TitleBoxColor"]=Color3.fromRGB(math.clamp(color.TitleBoxColor.R*255,0,255), math.clamp(color.TitleBoxColor.G*255,0,255), math.clamp(color.TitleBoxColor.B*255,0,255)),
	}
	return selectedColor
end
function SetColorDef()
	return colors.NILColor
end
function module.notify(target,title,text,color,dur)

	if target:WaitForChild("PlayerGui"):FindFirstChild("NotifyModuleContainer") ~= nil then

		local duration

		if dur == nil then
			duration = 3
		else if dur ~= nil and dur >= 0.1 then
				duration = tonumber(dur)
			end
		end
		--Rescripted by botwot7024 (Below this line)
		if title ~= nil and text ~= nil and color ~= nil then
			local selectedColor = colors.default
			if type(color)=="string" then--Only here for backwards compatibility(Updated to new color system)
				if string.lower(color) == "alert" or string.lower(color) == "success" or string.lower(color) == "warning" or string.lower(color) == "info" or string.lower(color) == "default" then
					if string.lower(color) == "alert" then
						selectedColor = colors.alert
					elseif string.lower(color) == "success" then
						selectedColor = colors.success
					elseif string.lower(color) == "warning" then
						selectedColor = colors.warning
					elseif string.lower(color) == "info" then
						selectedColor = colors.info
					elseif string.lower(color) == "default" then
						selectedColor = colors.default
					end
				else--if String but not table or know preset then default to NIL color
					warn("Color Select Was Missing, Defaulted to NIL color!")
					selectedColor = colors.NILColor
				end	
			elseif type(color)=="table" then--If new format then
				local Fail=false
				if color.R==nil or color.G==nil or color.B==nil then--The New DefaultColor System
					Fail=true
				elseif color.BackdropColor==nil or color.IconboxColor==nil or color.MSGBoxColor==nil or color.TitleBoxColor==nil then
					Fail=true
				else
					if color.BackdropColor.R==nil or color.BackdropColor.G==nil or color.BackdropColor.B==nil then
						Fail=true
					elseif color.IconboxColor.R==nil or color.IconboxColor.G==nil or color.IconboxColor.B==nil then
						Fail=true
					elseif color.MSGBoxColor.R==nil or color.MSGBoxColor.G==nil or color.MSGBoxColor.B==nil then
						Fail=true
					elseif color.TitleBoxColor.R==nil or color.TitleBoxColor.G==nil or color.TitleBoxColor.B==nil then
						Fail=true
					else
					 
						--Disable this and add in as a setting, This being on here means if the color is missing "G" but isint nil
						--then it will result in Off-looking Color schemes(Generally looks broken/Fucked up)
						--selectedColor = SetColorCus(color)
					end
				end
				if Fail then
					selectedColor = SetColorDef()
				else --Needed
					--Enabled here due to Else condition usage
					selectedColor = SetColorCus(color)
				end			
			else--if no color defined at all default to "Default"
				selectedColor = SetColorDef()
			end
			--Rescripted by botwot7024 (Above this line)
			local notification_number = 1


			if game.Players:FindFirstChild(target.Name) then
				local player = game.Players:FindFirstChild(target.Name)

				for _,v in pairs(player.PlayerGui.NotifyModuleContainer.OnScreenCache.Notifications:GetChildren()) do
					local S1=string.split(v.Name,"Notification")
					--print(S1)
					if S1[2]==nil then
						S1[2]=0
					end

Updater.lua
--Settings
local ExpirmentalUpdates=true
local UseCustomVersion=false

--Script Start
if UseCustomVersion then
	ExpirmentalUpdates=false
end
if ExpirmentalUpdates then
	UseCustomVersion=false
end
local TweenService = game.Workspace:WaitForChild("Services"):FindFirstChild("notify"):FindFirstChild("TweenService"):FindFirstChild("Event") -- Replace "notify" with the location of the module
local NVersion=script.Parent:WaitForChild("Version")
--Gather Services
local CodeService = game:GetService("HttpService")
local NewCode = CodeService:GetAsync("https://raw.githubusercontent.com/Coaxgames/NotifyModuleV1-Reworked/CommunityRelease/Changes.md")
--local NewInst = CodeService:GetAsync("https://raw.githubusercontent.com/Coaxgames/NotifyModuleV1-Reworked/CommunityRelease/InstuctionForModule.md")
--print(NewCode)
--Get local Notify Version
local Var1=script.Parent.Version.Value
local SplitV1=string.split(Var1, "V")
print(SplitV1)
local LocalVersion=SplitV1[2]
local FindVersion--GitVersion
LocalVersion=tonumber(LocalVersion)
if LocalVersion==nil then
	print("Cannot Update, Local version is missing")
	LocalVersion="Fallback"
end
--Look for version inside of NewCode to see if we are up to date
if NewCode~=nil then
	NewCode=string.gsub(NewCode, "```", "")
	FindVersion=string.match(NewCode, "NVersion.Value="..'%b""')
	FindVersion=string.gsub(FindVersion, "NVersion.Value=", "")
	FindVersion=string.gsub(FindVersion, "\"", "")
	FindVersion=string.gsub(FindVersion, "V", "")
	FindVersion=tonumber(FindVersion)
	print("Local Version="..LocalVersion..", GitVersion="..FindVersion)
	if LocalVersion~=nil and LocalVersion < FindVersion then
		if ExpirmentalUpdates then
			print("Requesting Live Patch...")
			if game:GetService("ServerScriptService"):FindFirstChild("NotifyUpdater"):FindFirstChild("UpdateComs") then
				game:GetService("ServerScriptService").NotifyUpdater.UpdateComs:Fire(script.Parent.Parent:FindFirstChild("notify"), LocalVersion, FindVersion, NewCode)
				return	
			else
				print("Cannot Patch: Missing UpdateComs Event")
			end
			print("Outdated Version, Please Update with NotifyUpdater...")
			warn("A new Update is available for NotifyV"..LocalVersion.." --> V"..FindVersion.."\nUse NotifyUpdater if you have it")
		else
			print("Updates are turned off, notify is out-of-date")
		end
	else
		print("Updates are turned on, notify is up-To-Date")
	end
end

--StartUp




GUIBuilder.lua
local module = {}


function module.CreateBin(Target)
	-- Instances:

	local NotifyModuleContainer = Instance.new("ScreenGui")
	local OffScreenCache = Instance.new("Folder")
	local Assets = Instance.new("Folder")
	local Presets = Instance.new("Folder")
	local SettingsButton = Instance.new("ImageButton")
	local UICorner_4 = Instance.new("UICorner")
	local Overlay = Instance.new("Frame")
	local UICorner_5 = Instance.new("UICorner")
	local ICON_2 = Instance.new("ImageLabel")
	local UICorner_6 = Instance.new("UICorner")
	local MSG_2 = Instance.new("TextBox")
	local UICorner_7 = Instance.new("UICorner")
	local Cache = Instance.new("Folder")
	local Input = Instance.new("TextBox")
	local UICorner_8 = Instance.new("UICorner")
	local PUserSettings = Instance.new("Folder")
	local OnScreenCache = Instance.new("Frame")
	local UIListLayout = Instance.new("UIListLayout")
	local Notifications = Instance.new("Folder")

	--Properties:

	NotifyModuleContainer.Name = "NotifyModuleContainer"
	NotifyModuleContainer.Parent = Target
	NotifyModuleContainer.ZIndexBehavior = Enum.ZIndexBehavior.Sibling

	OffScreenCache.Name = "OffScreenCache"
	OffScreenCache.Parent = NotifyModuleContainer

	Assets.Name = "Assets"
	Assets.Parent = NotifyModuleContainer

	Presets.Name = "Presets"
	Presets.Parent = Assets

	SettingsButton.Name = "SettingsButton"
	SettingsButton.Parent = NotifyModuleContainer
	SettingsButton.BackgroundColor3 = Color3.fromRGB(255, 255, 255)
	SettingsButton.BorderColor3 = Color3.fromRGB(0, 0, 0)
	SettingsButton.BorderSizePixel = 0
	SettingsButton.Position = UDim2.new(0.217289716, 0, 0.0181818176, 0)
	SettingsButton.Size = UDim2.new(0, 45, 0, 45)
	SettingsButton.Image = "rbxasset://textures/ui/GuiImagePlaceholder.png"

	UICorner_4.Parent = SettingsButton

	Overlay.Name = "Overlay"
	Overlay.Parent = SettingsButton
	Overlay.BackgroundColor3 = Color3.fromRGB(255, 255, 255)
	Overlay.BackgroundTransparency = 0.650
	Overlay.BorderColor3 = Color3.fromRGB(0, 0, 0)
	Overlay.BorderSizePixel = 0
	Overlay.Position = UDim2.new(-2.13217044, 0, 2.45861149, 0)
	Overlay.Size = UDim2.new(5.19883728, 0, -0.269083142, 80)
	Overlay.Visible = false

	UICorner_5.Parent = Overlay

	ICON_2.Name = "ICON"
	ICON_2.Parent = Overlay
	ICON_2.BackgroundColor3 = Color3.fromRGB(255, 255, 255)
	ICON_2.BorderColor3 = Color3.fromRGB(0, 0, 0)
	ICON_2.BorderSizePixel = 0
	ICON_2.Position = UDim2.new(0.0336336195, 0, 0.102357857, 0)
	ICON_2.Size = UDim2.new(0, 50, 0, 50)
	ICON_2.Image = "rbxasset://textures/ui/GuiImagePlaceholder.png"

	UICorner_6.Parent = ICON_2

	MSG_2.Name = "MSG"
	MSG_2.Parent = Overlay
	MSG_2.BackgroundColor3 = Color3.fromRGB(255, 0, 0)
	MSG_2.BorderColor3 = Color3.fromRGB(0, 0, 0)
	MSG_2.BorderSizePixel = 0
	MSG_2.Position = UDim2.new(0.275236875, 0, 0.106087409, 0)
	MSG_2.Size = UDim2.new(0, 163, 0, 50)
	MSG_2.Font = Enum.Font.SourceSans
	MSG_2.Text = ""
	MSG_2.TextColor3 = Color3.fromRGB(0, 0, 0)
	MSG_2.TextSize = 14.000

	UICorner_7.Parent = MSG_2

	Cache.Name = "Cache"
	Cache.Parent = SettingsButton

	Input.Name = "Input"
	Input.Parent = Cache
	Input.BackgroundColor3 = Color3.fromRGB(255, 0, 0)
	Input.BorderColor3 = Color3.fromRGB(0, 0, 0)
	Input.BorderSizePixel = 0
	Input.Position = UDim2.new(-1.32476294, 0, 1.15053189, 0)
	Input.Size = UDim2.new(0, 163, 0, 50)
	Input.Visible = false
	Input.Font = Enum.Font.SourceSans
	Input.Text = ""
	Input.TextColor3 = Color3.fromRGB(0, 0, 0)
	Input.TextSize = 14.000

	UICorner_8.Parent = Input

	PUserSettings.Name = "UserSettings"
	PUserSettings.Parent = NotifyModuleContainer

	OnScreenCache.Name = "OnScreenCache"
	OnScreenCache.Parent = NotifyModuleContainer
	OnScreenCache.BackgroundColor3 = Color3.fromRGB(255, 255, 255)
	OnScreenCache.BackgroundTransparency = 0.900
	OnScreenCache.BorderColor3 = Color3.fromRGB(0, 0, 0)
	OnScreenCache.BorderSizePixel = 0
	OnScreenCache.Position = UDim2.new(0.284658819, 0, 0.0181818213, 0)
	OnScreenCache.Size = UDim2.new(0.355598003, 0, 0.501652896, 0)
	
	Notifications.Name = "Notifications"
	Notifications.Parent = OnScreenCache
	
	UIListLayout.Parent = game.StarterGui.NotifyModuleContainer.OnScreenCache.Notifications
	UIListLayout.HorizontalAlignment = Enum.HorizontalAlignment.Center
	UIListLayout.VerticalAlignment = Enum.VerticalAlignment.Bottom
	
	
	
	--Create Notify Object
	local Notify1 = Instance.new("Frame")
	local UICorner = Instance.new("UICorner")
	local ICON = Instance.new("ImageLabel")
	local UICorner_2 = Instance.new("UICorner")
	local MSG = Instance.new("TextBox")
	local UICorner_3 = Instance.new("UICorner")
	local Title = Instance.new("TextBox")
	local UICorner_4 = Instance.new("UICorner")

	--Properties:

	Notify1.Name = "Notify1"
	Notify1.Parent = Assets
	Notify1.BackgroundColor3 = Color3.fromRGB(255, 255, 255)
	Notify1.BackgroundTransparency = 0.650
	Notify1.BorderColor3 = Color3.fromRGB(0, 0, 0)
	Notify1.BorderSizePixel = 0
	Notify1.Position = UDim2.new(0.050000038, 0, 0, 0)
	Notify1.Size = UDim2.new(0.899999976, 0, 0.300000012, 0)
	Notify1.Visible = false

	UICorner.Parent = Notify1

	ICON.Name = "ICON"
	ICON.Parent = Notify1
	ICON.BackgroundColor3 = Color3.fromRGB(255, 255, 255)
	ICON.BorderColor3 = Color3.fromRGB(0, 0, 0)
	ICON.BorderSizePixel = 0
	ICON.Position = UDim2.new(0.0268115904, 0, 0.293825477, 0)
	ICON.Size = UDim2.new(0.16695632, 0, 0.650000095, 0)
	ICON.Image = "rbxasset://textures/ui/GuiImagePlaceholder.png"

	UICorner_2.Parent = ICON

	MSG.Name = "MSG"
	MSG.Parent = Notify1
	MSG.BackgroundColor3 = Color3.fromRGB(255, 0, 0)
	MSG.BorderColor3 = Color3.fromRGB(0, 0, 0)
	MSG.BorderSizePixel = 0
	MSG.Position = UDim2.new(0.209000021, 0, 0.293825477, 0)
	MSG.Size = UDim2.new(0.769999981, 0, 0.649999976, 0)
	MSG.Font = Enum.Font.SourceSans
	MSG.Text = ""
	MSG.TextColor3 = Color3.fromRGB(0, 0, 0)
	MSG.TextScaled = true
	MSG.TextSize = 14.000
	MSG.TextWrapped = true

	UICorner_3.Parent = MSG

	Title.Name = "Title"
	Title.Parent = Notify1
	Title.BackgroundColor3 = Color3.fromRGB(255, 0, 0)
	Title.BorderColor3 = Color3.fromRGB(0, 0, 0)
	Title.BorderSizePixel = 0
	Title.Position = UDim2.new(0.0270000007, 0, 0.0250000004, 0)
	Title.Size = UDim2.new(0.952000022, 0, 0.25, 0)
	Title.Font = Enum.Font.SourceSans
	Title.Text = ""
	Title.TextColor3 = Color3.fromRGB(0, 0, 0)
	Title.TextScaled = true
	Title.TextSize = 14.000
	Title.TextWrapped = true

	UICorner_4.Parent = Title
	
	return NotifyModuleContainer

end



return module


UpdaterPlugin.lua
local ChangeHistoryService = game:GetService("ChangeHistoryService")
local Selection = game:GetService("Selection")

-- Create a new toolbar section titled "Custom Script Tools"
local toolbar = plugin:CreateToolbar("Custom Script Tools")


-- Add a toolbar button named "Create Empty Script"
local newScriptButton = toolbar:CreateButton("Debug", "Test Services", "rbxassetid://4458901886")
-- Make button clickable even if 3D viewport is hidden
newScriptButton.ClickableWhenViewportHidden = false

--UpdaterComs Path
local UpdaterComs=game:GetService("ServerScriptService"):WaitForChild("NotifyUpdater"):WaitForChild("UpdateComs")

UpdaterComs.Event:Connect(function(NotifyPath, LocalVersion, FindVersion, NewCode, NewInst)
	if LocalVersion < FindVersion then
		print("Debug")
		print(LocalVersion)
		print(FindVersion)
	end
	print("Updater: Updating From V"..LocalVersion..", --> V"..FindVersion)
	--Move Children to the Cache(in ServerScriptSerive)
	local SSS=game:GetService("ServerScriptService")
	if SSS:FindFirstChild("NotifyUpdater") then
		SSS=SSS:FindFirstChild("NotifyUpdater"):FindFirstChild("Cached")
	end
	for i, Obj in pairs(NotifyPath:GetChildren()) do
		print("Updater: Moving "..Obj.Name.." To Server Cache")
		Obj.Parent=SSS
	end
	print("Updater: Removing Notify V"..LocalVersion)
	NotifyPath.Name="notnotify"
	
	--Create new Module with updated code
	print("Updater: Creating Notify V"..FindVersion)
	local NewNotify=Instance.new("ModuleScript")
	NewNotify.Name="notify"
	for i2, Obj2 in pairs(SSS:GetChildren()) do
		print("Updater: Moving "..Obj2.Name.." from Cache To Notify V"..FindVersion)
		Obj2.Parent=NewNotify
		if Obj2.Name=="Version" then
			Obj2.Value="V"..FindVersion
			print("Updater: Updated Version Value")
		end
	end
	
	--Update the Contents of new notify to NewCode, Also update version
	print("Updater: Writing Notify V"..FindVersion.."'s SRC")
	NewCode=string.gsub(NewCode, "`", "")
	NewNotify.Source=NewCode
	NewNotify.Parent=NotifyPath.Parent
	print("Updater: Moving Notify V"..LocalVersion.." to Cache")
	NotifyPath.Parent=SSS
	print("Updater: Finished updating to V"..FindVersion)
end)



local function onNewScriptButtonClicked()
	--local selectedObjects = Selection:Get()
	--local newScript = Instance.new("Script")
	--newScript.Source = ""
	--newScript.Parent = parent
	--ChangeHistoryService:SetWaypoint("")
end


newScriptButton.Click:Connect(onNewScriptButtonClicked)


Changes.lua
local module = {}
local TweenService = game.Workspace:WaitForChild("Services"):FindFirstChild("notify"):FindFirstChild("TweenService"):FindFirstChild("Event") -- Replace "notify" with the location of the module
local GUIBuilder = require(script:WaitForChild("GUIBuilder"))
local NVersion=script:WaitForChild("Version")
NVersion.Value="V1.05"
print("Started notify Module")
function insertContainer(p)
	if p == nil then
		--Rewrite
		if game:GetService("StarterGui"):WaitForChild("NotifyModuleContainer")==nil then
			GUIBuilder.CreateBin(game:GetService("StarterGui"))
		end
	else
		--Rewrite
		if p:WaitForChild("PlayerGui"):WaitForChild("NotifyModuleContainer")==nil then
			GUIBuilder.CreateBin(p:FindFirstChild("PlayerGui"))
		end
	end
end

insertContainer()

--Preset colors
local colors = {
	NILColor  = {
		["BackdropColor"]=Color3.fromRGB(40, 40, 40),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(40, 40, 40),
	},
	alert  = {
		["BackdropColor"]=Color3.fromRGB(254, 104, 104),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(254, 104, 104),
	},
	success = {
		["BackdropColor"]=Color3.fromRGB(104, 254, 104),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(104, 254, 104),
	},
	warning  = {
		["BackdropColor"]=Color3.fromRGB(254, 254, 104),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(254, 254, 104),
	},
	info = {
		["BackdropColor"]=Color3.fromRGB(104, 104, 254),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(104, 104, 254),
	},
	default = {
		["BackdropColor"]=Color3.fromRGB(254, 254, 254),
		["IconboxColor"]=Color3.fromRGB(54, 54, 54),
		["MSGBoxColor"]=Color3.fromRGB(54, 54, 54),
		["TitleBoxColor"]=Color3.fromRGB(254, 254, 254),
	},
}

--Color functions
function SetColorCus(color)
	local selectedColor
	selectedColor = {
		["BackdropColor"]=Color3.fromRGB(math.clamp(color.BackdropColor.R*255,0,255), math.clamp(color.BackdropColor.G*255,0,255), math.clamp(color.BackdropColor.B*255,0,255)),
		["IconboxColor"]=Color3.fromRGB(math.clamp(color.IconboxColor.R*255,0,255), math.clamp(color.IconboxColor.G*255,0,255), math.clamp(color.IconboxColor.B*255,0,255)),
		["MSGBoxColor"]=Color3.fromRGB(math.clamp(color.MSGBoxColor.R*255,0,255), math.clamp(color.MSGBoxColor.G*255,0,255), math.clamp(color.MSGBoxColor.B*255,0,255)),
		["TitleBoxColor"]=Color3.fromRGB(math.clamp(color.TitleBoxColor.R*255,0,255), math.clamp(color.TitleBoxColor.G*255,0,255), math.clamp(color.TitleBoxColor.B*255,0,255)),
	}
	return selectedColor
end
function SetColorDef()
	return colors.NILColor
end
function module.notify(target,title,text,color,dur)

	if target:WaitForChild("PlayerGui"):FindFirstChild("NotifyModuleContainer") ~= nil then

		local duration

		if dur == nil then
			duration = 3
		else if dur ~= nil and dur >= 0.1 then
				duration = tonumber(dur)
			end
		end
		--Rescripted by botwot7024 (Below this line)
		if title ~= nil and text ~= nil and color ~= nil then
			local selectedColor = colors.default
			if type(color)=="string" then--Only here for backwards compatibility(Updated to new color system)
				if string.lower(color) == "alert" or string.lower(color) == "success" or string.lower(color) == "warning" or string.lower(color) == "info" or string.lower(color) == "default" then
					if string.lower(color) == "alert" then
						selectedColor = colors.alert
					elseif string.lower(color) == "success" then
						selectedColor = colors.success
					elseif string.lower(color) == "warning" then
						selectedColor = colors.warning
					elseif string.lower(color) == "info" then
						selectedColor = colors.info
					elseif string.lower(color) == "default" then
						selectedColor = colors.default
					end
				else--if String but not table or know preset then default to NIL color
					warn("Color Select Was Missing, Defaulted to NIL color!")
					selectedColor = colors.NILColor
				end	
			elseif type(color)=="table" then--If new format then
				local Fail=false
				if color.R==nil or color.G==nil or color.B==nil then--The New DefaultColor System
					Fail=true
				elseif color.BackdropColor==nil or color.IconboxColor==nil or color.MSGBoxColor==nil or color.TitleBoxColor==nil then
					Fail=true
				else
					if color.BackdropColor.R==nil or color.BackdropColor.G==nil or color.BackdropColor.B==nil then
						Fail=true
					elseif color.IconboxColor.R==nil or color.IconboxColor.G==nil or color.IconboxColor.B==nil then
						Fail=true
					elseif color.MSGBoxColor.R==nil or color.MSGBoxColor.G==nil or color.MSGBoxColor.B==nil then
						Fail=true
					elseif color.TitleBoxColor.R==nil or color.TitleBoxColor.G==nil or color.TitleBoxColor.B==nil then
						Fail=true
					else
					 
						--Disable this and add in as a setting, This being on here means if the color is missing "G" but isint nil
						--then it will result in Off-looking Color schemes(Generally looks broken/Fucked up)
						--selectedColor = SetColorCus(color)
					end
				end
				if Fail then
					selectedColor = SetColorDef()
				else --Needed
					--Enabled here due to Else condition usage
					selectedColor = SetColorCus(color)
				end			
			else--if no color defined at all default to "Default"
				selectedColor = SetColorDef()
			end
			--Rescripted by botwot7024 (Above this line)
			local notification_number = 1


			if game.Players:FindFirstChild(target.Name) then
				local player = game.Players:FindFirstChild(target.Name)

				for _,v in pairs(player.PlayerGui.NotifyModuleContainer.OnScreenCache.Notifications:GetChildren()) do
					local S1=string.split(v.Name,"Notification")
					--print(S1)
					if S1[2]==nil then
						S1[2]=0
					end
					notification_number = tonumber(S1[2])+((notification_number+1)/100)
				end
				
				
				--READ ME | DEV NOTE
				--Change this to a server Asset, Rather than creating it here.
				--[[
				
					New Color format: (Still thinking on this)
					Color={
						["BackdropColor"]={
							["R"]=255, --Clamped To 0-255
							["G"]=0, --Clamped To 0-255
							["B"]=0, --Clamped To 0-255	
						},
						["IconboxColor"]={
							["R"]=255, --Clamped To 0-255
							["G"]=0, --Clamped To 0-255
							["B"]=0, --Clamped To 0-255	
						},
						["MSGBoxColor"]={
							["R"]=255, --Clamped To 0-255
							["G"]=0, --Clamped To 0-255
							["B"]=0, --Clamped To 0-255	
						},
						["TitleBoxColor"]={
							["R"]=255, --Clamped To 0-255
							["G"]=0, --Clamped To 0-255
							["B"]=0, --Clamped To 0-255	
						},
					}
				]]
				--Rewrite
				local NewNotification= player.PlayerGui:WaitForChild("NotifyModuleContainer").Assets.Notify1:Clone()
				NewNotification.Parent=player.PlayerGui.NotifyModuleContainer.OnScreenCache.Notifications
				NewNotification.Name="Notification"..notification_number
				NewNotification.BackgroundColor3 = selectedColor.BackdropColor
				NewNotification.Visible=true
				
				NewNotification.Title.Text=title
				NewNotification.Title.BackgroundColor3 = selectedColor.TitleBoxColor
				NewNotification.Title.TextScaled=true
				NewNotification.Title.TextWrapped=true
				
				NewNotification.MSG.Text=text
				NewNotification.MSG.BackgroundColor3 = selectedColor.MSGBoxColor
				NewNotification.MSG.TextScaled=true
				NewNotification.MSG.TextWrapped=true
				
				
				NewNotification.ICON.Image="rbxasset://textures/ui/GuiImagePlaceholder.png"--[[icon]]
				NewNotification.ICON.BackgroundColor3 = selectedColor.IconboxColor
			

				--Rescripted by botwot7024
				if script:FindFirstChild("TweenService") then
					local Time=os.clock()
					--print("Tweening 1 at "..Time)
					TweenService:Fire(NewNotification, UDim2.new(0.95, 0, -0.014, 80), duration, UDim2.new(0.95, 0, 0, 0), 1.1, Time)
					--print("Tweening 2 at "..os.clock())
				else
					NewNotification:TweenSize(UDim2.new(0.95, 0, -0.014, 80))
					wait(duration)
					
					NewNotification:TweenSize(UDim2.new(0, 0, 0, 0))
					wait(1.1)
					
					NewNotification:Destroy()
				end
				--Rescripted by botwot7024
			end
		end
	else
		insertContainer(target)
		wait(0.05)
		module.notify(target,title,text,color,dur)
	end

end

return module
